File: Week 15/java/1.HeapAndPriorityQueue.java
/*
 * WEEK 15 - HEAPS & PRIORITY QUEUES
 * Topic: Heap, Heap Sort, Priority Queue Applications
 *
 * HEAP: A complete binary tree stored as an array where:
 * - MAX-HEAP: every parent >= children (root = max element)
 * - MIN-HEAP: every parent <= children (root = min element)
 *
 * ARRAY REPRESENTATION (1-indexed for clarity):
 * - Parent of node i: i / 2
 * - Left child of i: 2 * i
 * - Right child of i: 2 * i + 1
 *
 * OPERATIONS:
 * - insert:    Add at end, sift UP     O(log n)
 * - extractMax/Min: Replace root with last, sift DOWN  O(log n)
 * - buildHeap: Convert array to heap   O(n) — better than n inserts!
 *              (Start from last non-leaf, sift down each)
 * - peekMax:   O(1) — root element
 *
 * HEAP SORT:
 * 1. buildHeap O(n)
 * 2. Extract max n times, place at end O(n log n)
 * Overall: O(n log n), in-place, O(1) space
 * NOT stable.
 *
 * CLASSIC PROBLEMS:
 * - Kth largest element: O(n log k) using min-heap of size k
 * - Merge K sorted arrays: O(n log k) using min-heap
 * - Median from stream: O(log n) per insert using two heaps
 */

import java.util.*;

public class HeapAndPriorityQueue {

    // --- MAX-HEAP from scratch (0-indexed array) ---
    static class MaxHeap {
        int[] data;
        int size;

        MaxHeap(int capacity) { data = new int[capacity]; size = 0; }

        int parent(int i) { return (i - 1) / 2; }
        int left(int i) { return 2 * i + 1; }
        int right(int i) { return 2 * i + 2; }

        void swap(int i, int j) { int t = data[i]; data[i] = data[j]; data[j] = t; }

        // Sift up: fix heap property from node i upward
        void siftUp(int i) {
            while (i > 0 && data[parent(i)] < data[i]) {
                swap(i, parent(i));
                i = parent(i);
            }
        }

        // Sift down: fix heap property from node i downward
        void siftDown(int i) {
            int largest = i;
            int l = left(i), r = right(i);
            if (l < size && data[l] > data[largest]) largest = l;
            if (r < size && data[r] > data[largest]) largest = r;
            if (largest != i) { swap(i, largest); siftDown(largest); }
        }

        void insert(int x) { data[size++] = x; siftUp(size - 1); }
        int extractMax() {
            int max = data[0];
            data[0] = data[--size];
            siftDown(0);
            return max;
        }
        int peekMax() { return data[0]; }
        boolean isEmpty() { return size == 0; }
    }

    // Heap Sort using max-heap (in-place)
    static void heapSort(int[] arr) {
        int n = arr.length;
        // Build max-heap: start from last non-leaf, sift down
        for (int i = n / 2 - 1; i >= 0; i--) siftDownInPlace(arr, n, i);
        // Extract max one by one: place at end
        for (int i = n - 1; i > 0; i--) {
            int t = arr[0]; arr[0] = arr[i]; arr[i] = t; // swap max to end
            siftDownInPlace(arr, i, 0); // restore heap for reduced size
        }
    }
    static void siftDownInPlace(int[] arr, int n, int i) {
        int largest = i, l = 2*i+1, r = 2*i+2;
        if (l < n && arr[l] > arr[largest]) largest = l;
        if (r < n && arr[r] > arr[largest]) largest = r;
        if (largest != i) {
            int t = arr[i]; arr[i] = arr[largest]; arr[largest] = t;
            siftDownInPlace(arr, n, largest);
        }
    }

    // Kth Largest Element — min-heap of size k
    // Time: O(n log k), Space: O(k)
    static int kthLargest(int[] arr, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>(); // default = min-heap
        for (int x : arr) {
            minHeap.offer(x);
            if (minHeap.size() > k) minHeap.poll(); // remove smallest
        }
        return minHeap.peek(); // kth largest
    }

    // Median from a Data Stream — two heaps
    // maxHeap (left half) | minHeap (right half)
    // Maintain: maxHeap.size() == minHeap.size() OR maxHeap.size() == minHeap.size() + 1
    static class MedianFinder {
        PriorityQueue<Integer> lower = new PriorityQueue<>(Collections.reverseOrder()); // max-heap for lower half
        PriorityQueue<Integer> upper = new PriorityQueue<>(); // min-heap for upper half

        void addNum(int num) {
            lower.offer(num);
            upper.offer(lower.poll()); // ensure all upper > all lower
            if (upper.size() > lower.size()) lower.offer(upper.poll()); // balance sizes
        }

        double findMedian() {
            if (lower.size() > upper.size()) return lower.peek();
            return (lower.peek() + upper.peek()) / 2.0;
        }
    }

    public static void main(String[] args) {
        // Max-Heap
        System.out.println("=== Max-Heap ===");
        MaxHeap heap = new MaxHeap(20);
        int[] nums = {5, 3, 7, 1, 9, 2, 8};
        for (int x : nums) heap.insert(x);
        System.out.print("Extract in order: ");
        while (!heap.isEmpty()) System.out.print(heap.extractMax() + " "); // 9 8 7 5 3 2 1
        System.out.println();

        // Heap Sort
        System.out.println("\n=== Heap Sort ===");
        int[] arr = {12, 11, 13, 5, 6, 7};
        System.out.println("Before: " + Arrays.toString(arr));
        heapSort(arr);
        System.out.println("After:  " + Arrays.toString(arr));

        // Kth Largest
        System.out.println("\n=== Kth Largest ===");
        int[] arr2 = {3, 2, 1, 5, 6, 4};
        for (int k = 1; k <= arr2.length; k++)
            System.out.println("k=" + k + " → " + kthLargest(arr2, k));

        // Java PriorityQueue (min-heap by default)
        System.out.println("\n=== Java PriorityQueue ===");
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int x : new int[]{5, 1, 3, 2, 4}) pq.offer(x);
        System.out.print("Poll order (min-heap): ");
        while (!pq.isEmpty()) System.out.print(pq.poll() + " ");
        System.out.println();

        // Median from stream
        System.out.println("\n=== Median from Stream ===");
        MedianFinder mf = new MedianFinder();
        int[] stream = {5, 15, 1, 3, 2, 8, 7, 9, 10, 6, 11, 4};
        for (int x : stream) {
            mf.addNum(x);
            System.out.printf("Added %2d → median = %.1f%n", x, mf.findMedian());
        }
    }
}

File: Week 15/python/1.HeapAndPriorityQueue.py
"""
WEEK 15 - PYTHON DSA
Topic: Heap, Heap Sort, Priority Queue Applications
File: 1.HeapAndPriorityQueue.py

CONCEPT:
A heap is a complete binary tree stored as an array satisfying the heap
property:
  - Max-heap: every parent >= children (root is max).
  - Min-heap: every parent <= children (root is min).
Stored 0-indexed: parent(i) = (i-1)//2; left = 2i+1; right = 2i+2.

KEY POINTS:
- insert: append at end, sift up. O(log n).
- extractMax/Min: swap root with last, shrink, sift down. O(log n).
- buildHeap: O(n) — start from last non-leaf, sift down each. Better than n
  inserts (O(n log n)).
- Heap sort: build max-heap, repeatedly extract max to the end. O(n log n).

ALGORITHM / APPROACH:
- kth largest: maintain a min-heap of size k; the root is the kth largest.
- merge K sorted arrays: min-heap of (value, array_index, element_index).
- median of stream: two heaps. Max-heap holds lower half; min-heap holds
  upper half. Insert into max-heap then push max into min-heap; rebalance
  sizes so max-heap has equal-or-one-more.

PYTHON-SPECIFIC NOTES:
- The standard library `heapq` is a MIN-heap only. To build a max-heap we
  negate values (push -x, pop -x). heapq operates on a Python list.
- heapq.heappush / heappop are O(log n). heapify is O(n).
- For full priority queue API see queue.PriorityQueue (thread-safe).

DRY RUN:
Example 1: build max-heap from [5,3,7,1,9,2,8]
  Inserting 5: [5]
  Insert 3:    [5,3]
  Insert 7:    [5,3,7] -> sift 7 up: [7,3,5]
  Insert 1:    [7,3,5,1]
  Insert 9:    [7,3,5,1,9] -> sift 9 up to root: [9,7,5,1,3]
  Insert 2:    [9,7,5,1,3,2]
  Insert 8:    [9,7,5,1,3,2,8] -> sift 8: [9,7,8,1,3,2,5]
  extractMax order: 9 8 7 5 3 2 1

Example 2: kthLargest([3,2,1,5,6,4], k=2) using min-heap of size 2
  Push 3 -> [3]
  Push 2 -> [2,3]
  Push 1 -> [1,2,3] size>k -> pop 1 -> [2,3]
  Push 5 -> [2,3,5] size>k -> pop 2 -> [3,5]
  Push 6 -> [3,5,6] -> pop 3 -> [5,6]
  Push 4 -> [4,5,6] -> pop 4 -> [5,6]
  Root = 5 = 2nd largest.

COMPLEXITY:
  insert/extract:   O(log n)
  buildHeap:        O(n)
  heapSort:         O(n log n) time, O(1) extra
  kthLargest:       O(n log k)
  median stream:    O(log n) per add, O(1) findMedian
"""

from __future__ import annotations
import heapq
from typing import List


class MaxHeap:
    """Hand-rolled max-heap on a Python list (0-indexed)."""

    def __init__(self) -> None:
        self.data: List[int] = []

    def __len__(self) -> int: return len(self.data)
    def is_empty(self) -> bool: return not self.data
    def peek_max(self) -> int: return self.data[0]

    @staticmethod
    def _parent(i: int) -> int: return (i - 1) // 2
    @staticmethod
    def _left(i: int) -> int: return 2 * i + 1
    @staticmethod
    def _right(i: int) -> int: return 2 * i + 2

    def _sift_up(self, i: int) -> None:
        while i > 0 and self.data[self._parent(i)] < self.data[i]:
            p = self._parent(i)
            self.data[i], self.data[p] = self.data[p], self.data[i]
            i = p

    def _sift_down(self, i: int) -> None:
        n = len(self.data)
        while True:
            l, r = self._left(i), self._right(i)
            largest = i
            if l < n and self.data[l] > self.data[largest]: largest = l
            if r < n and self.data[r] > self.data[largest]: largest = r
            if largest == i: return
            self.data[i], self.data[largest] = self.data[largest], self.data[i]
            i = largest

    def insert(self, x: int) -> None:
        self.data.append(x)
        self._sift_up(len(self.data) - 1)

    def extract_max(self) -> int:
        if not self.data:
            raise IndexError("Heap empty")
        top = self.data[0]
        last = self.data.pop()
        if self.data:
            self.data[0] = last
            self._sift_down(0)
        return top


def heap_sort(arr: List[int]) -> None:
    """In-place ascending sort using max-heap construction."""
    n = len(arr)

    def sift_down(start: int, end: int) -> None:
        i = start
        while True:
            l, r = 2*i + 1, 2*i + 2
            largest = i
            if l < end and arr[l] > arr[largest]: largest = l
            if r < end and arr[r] > arr[largest]: largest = r
            if largest == i: return
            arr[i], arr[largest] = arr[largest], arr[i]
            i = largest

    for i in range(n // 2 - 1, -1, -1):
        sift_down(i, n)
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        sift_down(0, i)


def kth_largest(arr: List[int], k: int) -> int:
    """Maintain a min-heap of size k; root is the kth largest."""
    h: List[int] = []
    for x in arr:
        heapq.heappush(h, x)
        if len(h) > k:
            heapq.heappop(h)
    return h[0]


class MedianFinder:
    """Two heaps: lower (max-heap via negation) | upper (min-heap)."""

    def __init__(self) -> None:
        self.lower: List[int] = []  # store as negatives for max-heap behavior
        self.upper: List[int] = []

    def add_num(self, num: int) -> None:
        # 1) push to lower (max-heap) — use -num
        heapq.heappush(self.lower, -num)
        # 2) move the top of lower to upper to maintain ordering
        heapq.heappush(self.upper, -heapq.heappop(self.lower))
        # 3) rebalance sizes so len(lower) >= len(upper)
        if len(self.upper) > len(self.lower):
            heapq.heappush(self.lower, -heapq.heappop(self.upper))

    def find_median(self) -> float:
        if len(self.lower) > len(self.upper):
            return float(-self.lower[0])
        return (-self.lower[0] + self.upper[0]) / 2.0


def main() -> None:
    print("=== Max-Heap ===")
    heap = MaxHeap()
    for x in [5, 3, 7, 1, 9, 2, 8]:
        heap.insert(x)
    out = []
    while not heap.is_empty():
        out.append(heap.extract_max())
    print("Extract in order:", out)  # 9 8 7 5 3 2 1

    print("\n=== Heap Sort ===")
    arr = [12, 11, 13, 5, 6, 7]
    print("Before:", arr)
    heap_sort(arr)
    print("After: ", arr)

    print("\n=== Kth Largest ===")
    arr2 = [3, 2, 1, 5, 6, 4]
    for k in range(1, len(arr2) + 1):
        print(f"k={k} -> {kth_largest(arr2, k)}")

    print("\n=== heapq min-heap demo ===")
    h: List[int] = []
    for x in [5, 1, 3, 2, 4]: heapq.heappush(h, x)
    poll = []
    while h: poll.append(heapq.heappop(h))
    print("Pop order:", poll)  # 1 2 3 4 5

    print("\n=== Median from Stream ===")
    mf = MedianFinder()
    for x in [5, 15, 1, 3, 2, 8, 7, 9, 10, 6, 11, 4]:
        mf.add_num(x)
        print(f"Added {x:2d} -> median = {mf.find_median():.1f}")


if __name__ == "__main__":
    main()


"""
NOTES (vs. Java):
- Java PriorityQueue is a MIN-heap by default; pass Collections.reverseOrder()
  for a max-heap. Python heapq is min-only — negate values for max-heap.
- Java's PriorityQueue<int[]> supports custom Comparator; Python's heapq
  compares tuples lexicographically, so (priority, payload) works naturally.
- buildHeap and sift logic is identical conceptually across languages.
"""
